tmk_core/common/chibios/eeprom_stm32_l4.c

```diff
@@ -0,0 +1,553 @@
+        return 0;
+    }
+
+    /* keep DataBuf cache in sync */
+    *(uint32_t *)&DataBuf[Address] = DataDWord;
+
+    /* perform the write into flash memory */
+    /* First, attempt to write directly into the compacted flash area */
+    FLASH_Status status = eeprom_write_direct_entry(Address);
+
+    if (!status) {
+        eeprom_write_log_dword_entry(Address);
+    }
+
+    return status;
+}
+
+uint8_t EEPROM_ReadDataByte(uint16_t Address) {
+    uint8_t DataByte = 0xFF;
+
+    if (Address < FEE_DENSITY_BYTES) {
+        DataByte = DataBuf[Address];
+    }
+
+    return DataByte;
+}
+
+uint16_t EEPROM_ReadDataWord(uint16_t Address) {
+    uint16_t DataWord = 0xFFFF;
+
+    if (Address < FEE_DENSITY_BYTES - 1) {
+        /* Check word alignment */
+        if (Address % 2) {
+            DataWord = DataBuf[Address] | (DataBuf[Address + 1] << 8);
+        } else {
+            DataWord = *(uint16_t *)(&DataBuf[Address]);
+        }
+    }
+
+    return DataWord;
+}
+
+/*****************************************************************************
+ *  Bind to eeprom_driver.c
+ *******************************************************************************/
+void eeprom_driver_init(void) { EEPROM_Init(); }
+
+void eeprom_driver_erase(void) { EEPROM_Erase(); }
+
+void eeprom_read_block(void *buf, const void *addr, size_t len) {
+    const uint8_t *src  = (const uint8_t *)addr;
+    uint8_t *      dest = (uint8_t *)buf;
+
+    /* Check word alignment */
+    if (len && (uint32_t)src % 2) {
+        /* Read the unaligned first byte */
+        *dest++ = EEPROM_ReadDataByte((const uintptr_t)((uint16_t *)src));
+        --len;
+    }
+
+    uint16_t value;
+    bool     aligned = ((uint32_t)dest % 2 == 0);
+    while (len > 1) {
+        value = EEPROM_ReadDataWord((const uintptr_t)((uint16_t *)src));
+        if (aligned) {
+            *(uint16_t *)dest = value;
+            dest += 2;
+        } else {
+            *dest++ = value;
+            *dest++ = value >> 8;
+        }
+        src += 2;
+        len -= 2;
+    }
+    if (len) {
+        *dest = EEPROM_ReadDataByte((const uintptr_t)src);
+    }
+}
+
+void eeprom_write_block(const void *buf, void *addr, size_t len) {
+    uint8_t *      dest = (uint8_t *)addr;
+    const uint8_t *src  = (const uint8_t *)buf;
+
+    /* Check word alignment */
+    if (len && (uintptr_t)dest % 2) {
+        /* Write the unaligned first byte */
+        EEPROM_WriteDataByte((uintptr_t)dest++, *src++);
+        --len;
+    }
+
+    if ((uintptr_t)dest % 4 == 0) {
+        uint32_t dwvalue;
+        bool     dwaligned = ((uint32_t)src % 4 == 0);
+
+        while (len > 1) {
+            if (dwaligned) {
+                dwvalue = *(uint32_t *)src;
+            } else {
+                dwvalue = *(uint8_t *)src | (*(uint8_t *)(src + 1) << 8) | (*(uint8_t *)(src + 2) << 16) | (*(uint8_t *)(src + 3) << 24);
+            }
+            EEPROM_WriteDataDWord((uintptr_t)((uint16_t *)dest), dwvalue);
```

Comment:
In theory you may add `EEPROM_WriteDataQWord()` (which would be either a 64-bit direct write or two 32-bit log writes) to handle the case of a large write into an empty EEPROM; however, I'm not sure whether such writes actually happen and are worth optimizing for.

Reply:
`EEPROM_WriteDataDWord` actually write 64 bit since word is 32 bit in STM32 MCU.